Resource Limits

Blub opened the pull request:
Currently containers inherit the resource limits of the process starting them, but it can be useful to configure them explicitly.
If for instance a container can be started via a system service or manually via `lxc-start` or some wrapper, these 2 ways of starting the container can end up giving it different limits. Both have the option to change the limits before starting, but if they do not have `CAP_SYS_RESOURCE` or are unprivileged they won't be able to raise them back up if they lowered them and would have to resort to forking first.

Limits which aren't explicitly specified will be inherited as before, so the default behavior stays the same.
As usual the last specified value for a limit takes precedence, specifying an empty value deletes it (iow. declares it as to be inherited as before).

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

Blub commented on src/lxc/conf.c:
```diff
@@ -4037,6 +4147,27 @@ int lxc_clear_cgroups(struct lxc_conf *c, const char *key)
+{
+	struct lxc_list *it, *next;
+	bool all = false;
+	const char *k = key + 10;
```
Yeah could make it less magical (note the access to `k` later depends on `all` being `false`, which is an easily overlooked dependency, and error prone). I should probably also change `lxc_clear_cgroups` where I took the pattern from ;-).

brauner commented:
jenkins: test this please

stgraber commented:
jenkins: test this please

Blub commented:
Squashed the above mentioned change into it: https://github.com/Blub/lxc/commit/be95f354d4e764d4ebed87240e84e9cadcad4623
And added a 4th commit doing the same to the other functions of this pattern.

stgraber commented:
Looks like we dropped the ball on this one... Any chance you can rebase on master and then we can get some reviews done and merge this?

I agree having this feature would be pretty useful so it'd be nice to have this included.

brauner commented:
@Blub, once you've rebased I'm more than happy to merge. (If you happen to have time to add some sort of minimal tests for this than ++ if not I understand. :))

mmxe commented:
I spent a day and a half fighting this problem. If you can push this to ubuntu updates asap it's greatly appreciated.

Blub commented:
I can revisit this next week I think.

Blub commented:
Rebased and did the proposed changes. (Changes can be viewed separately here: https://github.com/Blub/lxc/commits/limits-3 as the `fixup:...` commits. (Not sure if the last one should be a separate commit here as well?))

brauner commented:
jenkins: test this please

brauner commented:
@Blub, thanks for implementing this! :)
